### PR TITLE
Rewrite lora manager with hooks

### PR DESCRIPTION
**Pull request by neecapp**
@jordanramstad I usually don't wholesale rewrite things without talking to the person directly, so I apologize for that.

This implementation is still not perfect and performance dips when specifying a LoRA. I did not have time to check against other implementations, so there may be similar performance issues with them as well.

Feel free to use this code as you see fit, hope it helps.

### CODE REVIEW

File: ldm/modules/lora_manager.py [hunk 1]
```diff
@@ -1,245 +1,230 @@
...
 
     def load_lora_from_prompt(self, prompt: str):
+        self.applied_loras = {}
         for m in re.findall(self.lora_match, prompt):
             self.apply_lora_model(m)
 
     def load_lora(self):
         self.load_lora_from_prompt(self.prompt)
 
+    def unload_lora(self, lora_name: str):
```
Comment (neecapp) [inline ldm/modules/lora_manager.py:211]: This doesn't actually do anything and can be removed.

File: ldm/modules/lora_manager.py [hunk 2]
```diff
@@ -1,245 +1,230 @@
...
-            self.patch_module("text_encoder", processed_lora, module, name, child_module, scale)
+        self.unet = pipe.unet
+        self.text_encoder = pipe.text_encoder
+        self.device = torch.device(choose_torch_device())
+        self.dtype = pipe.unet.dtype
+        self.loras = {}
+        self.applied_loras = {}
+        self.hooks = []
+
+        def find_modules(prefix, root_module: torch.nn.Module, target_replace_modules) -> dict[str, torch.nn.Module]:
```
Comment (neecapp) [inline ldm/modules/lora_manager.py:134]: Taken from: https://github.com/kohya-ss/sd-scripts/blob/main/networks/merge_lora.py#L48

File: ldm/modules/lora_manager.py [hunk 3]
```diff
@@ -1,245 +1,230 @@
...
+    layers: dict[str, LoRALayer]
+    multiplier: float
+
+    def __init__(self, name: str, multiplier=1.0):
+        self.name = name
+        self.layers = {}
+        self.multiplier = multiplier
+
+
+UNET_TARGET_REPLACE_MODULE = ["Transformer2DModel", "Attention"]
```
Comment (neecapp) [inline ldm/modules/lora_manager.py:33]: From https://github.com/kohya-ss/sd-scripts/blob/main/networks/lora.py#L90

**Comment by jordanramstad** [comment]
rewriting it is fine @neecapp I have been doing that a few times now, testing different implementations to see what works. :P

**Comment by jordanramstad** [comment]
applied the work from @neecapp great work!

Its having some issues with burning still, but getting much closer now.

**Pull request merged**